docs/design/velero-krm-enhancement.md

```diff
@@ -0,0 +1,103 @@
+- User will be able to pass all resource list in velero plugin interface as compared to the current one to one mapping of resources 
+
+Currently openshift velero plugin interface can accept a single input in the input parameter for the given code snippet. A new plugin type will be added which will accept the input as a list of resources and the modified snippet will look like
+```
+func (p *RestorePlugin) Execute(input *velero.RestoreItemActionExecuteInput) 
+(*velero.RestoreItemActionExecuteOutput, error) {
+
+	// KRM function logic to transform a particular resource
+        // Convert input to KubeObject for resource transformation 
+
+}
+```
+Here input will be a list of resources, output will be tranformed resource and the function will consist of KRM logic
+
+ 
+## Design and Implementation Details
+
+The idea is divided into 2 phases.
+
+### Phase 1 
+It will consist of initially making changes in current Infrastructure so that parity can be achieved with BackupActionItemType and RestoreActionItemType and prove that we can use KRM functions with Velero Plugins.Further Phase 1 is divided into two parts.
+	 
+ - First Part will consist of addition of KRM function logic in existing code where there is a  single resource being passed to the function. `ParseKubeObject()` can be used to covert the input from Velero Backup/Restore plugin in form of KubeObject and further operations can be performed on KubeObject. In this way we can assess if KRM functions work well with velero OADP Plugin. First part is meant to prove that we can execute a KRM function from within an existing BackupItemAction plugin.
+
+![Velero OADP Existing flow design](../images/velero-exisitng-addon.png)
+
+
+- Second  Part will be development of a new plugin type `BackupItemsWithKRM` and `RestoreItemsWithKRM` that will allow a plugin author to take in a list of resources, execute KRM functions, and return a list of items with translated data to be included in the backup.
+
+
+![Velero OADP New Interface design](../images/velero-new-design.png)
+
+
+### Phase 2 
+ Modify the velero controller codebase to pass in to this plugin type a list of resources so that we achieve performance enhancement by reducing the amount of plugin calls.
+
+Example: 
+
+Let us consider restoring N routes. In the current scenario we have to call the plugin N number of time for restoring routes. But if we modify the codebase and plugin interface to pass list of resources to plugin this will only require a single gRPC call thus drastically imptroving the performance. 
+
+
+### Drawbacks
+
+The current solution has basic logic for implementation of plugins and replacing that with KRM functions will have two drawbacks
+
+- Adding a new velero plugin type for KRM functions will require significant overhaul of the velero server framework. This is likely something that will take extensive upstream discussion with the Velero community.
+- Rewriting the OADP controller code will take time and efforts.
```

Comment:
I don't see where are we rewriting the oadp controller can you point this out in the phases?